### PR TITLE
E2E testing in CI

### LINKED ISSUE #87: Add end-to-end testing to CircleCI pipeline
To better know if PRs add bugs, we should add the existing end-to-end testing to the CircleCI-Pipeline.

This requires:

* Configuring the ssh keypair name at the cloudprovider with a random prefix
* Deleting the ssh keypair before ending the e2e test
* Adding the `test-e2e` target to circleci

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

Add end2end testing to CI pipeline

**Which issue(s) this PR fixes** :
Will eventually resolve #87 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by alvaroaleman** [comment]
This won't clean up a testnode in case it was created and an error happened afterwards, however I think if this ever happens its fine to manually clean it up.

File: pkg/cloudprovider/provider/hetzner/provider.go [hunk 1]
```diff
@@ -133,38 +131,6 @@ func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
 	return nil
 }
 
-func ensureSSHKeysExist(ctx context.Context, client hcloud.SSHKeyClient, key *machinessh.PrivateKey) (*hcloud.SSHKey, error) {
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/hetzner/provider.go]: see my points in https://github.com/kubermatic/machine-controller/issues/107

**Comment by mrIncompetent** [comment]
Blocked by https://github.com/kubermatic/machine-controller/issues/119

**Comment by alvaroaleman** [comment]
IMHO this is done, PTAL @mrIncompetent

File: .circleci/config.yml [hunk 2]
```diff
@@ -62,6 +62,77 @@ jobs:
...
+      - restore_cache:
+          key: vendor-{{ checksum "Gopkg.lock" }}
+      - setup_remote_docker
+      - run: apk update && apk add make bash git
+      - run: |
+          set -e
+          docker create -v /go/src/github.com/kubermatic/machine-controller --name vol-container alpine /bin/true
+          docker cp . vol-container:/go/src/github.com/kubermatic/machine-controller/
+          export USE_VOLUME_CONTAINER=true && cd test/tools/verify && make verify
+          docker cp vol-container:/go/src/github.com/kubermatic/machine-controller/test/tools/verify/verify .
```
Comment (mrIncompetent) [inline .circleci/config.yml]: Can you quickly explain why we do this instead of a a simple `docker build` ?

**Comment by mrIncompetent** [comment]
I'm not a big fan of using docker container in the makefile in general.
It leads to overly complex Makefiles & build-files.

I would rather prefer the way Kubernetes is doing it. They have a wrapper script `run.sh` which is responsible for running make commands in a special docker container.
So you do `./hack/run.sh make`

File: .circleci/config.yml (hunk 2 above)
Comment (alvaroaleman) [reply .circleci/config.yml]: We could do a docker build but then we would have to push it on every build

**Comment by alvaroaleman** [comment]
IMHO it won't get any simpler if you put the glue code in a bash script instead of a Makefile. Also if you do it like K8s you always create the container, then decide if you actually need it.

**Comment by mrIncompetent** [comment]
As we use Alpine, creating a Container does not hurt.
The performance improvement we get, does not justify the overall script complexity IMO.

@guusvw @metalmatze @p0lyn0mial Some opinions here?

**Comment by p0lyn0mial** [comment]
I'd prefer to treat CI as a tool that makes calls to our code. By doing so source code is outside of a tool that can be potentially replaced in the future. We can also share the common code. If one wants to change something it doesn't have to dive in CI specific code.

**Comment by alvaroaleman** [comment]
> As we use Alpine, creating a Container does not hurt.

That is probably true

>The performance improvement we get, does not justify the overall script complexity IMO.

The complexity has nothing to do with using or not using `make` - If we moved that to a bashscipt, we would have the same complexity but in a different place. The advantage of make is that it offers a nicer interface for end users.

**Comment by mrIncompetent** [comment]
The complexity was targetet at the `docker cp`

**Comment by mrIncompetent** [comment]
Sorry for the delay!

**Comment by p0lyn0mial** [comment]
yay! we have e2e tests :)

**Pull request merged**